### ISSUE TITLE
Separate samples from sources and from application

### ISSUE THREAD

**Issue by lanwen** [swagger-api/swagger-codegen#537]
Hello!
I want to help write some unit tests and simplify complexity of the code.

But it seems that a lot of stuff here that should be in other place: 
- Samples: seems that they should be in separate repo called `swagger-codegen-samples` and simply use latest release of CLI-interface to build new versions.
- Swagger-generator: think it should be in repo called `swagger-generator` and contains modules:
  - swagger-web-app (with content of current module swagger-generator - is it same as in repo swagger-generator?)
  - swagger-cli (with content of current class [Codegen](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/java/com/wordnik/swagger/codegen/Codegen.java))

Also, think repo `swagger-parser` should be in this repo, because of parsing schema with classes from swagger-parser repo  only used to generate some kind of code.

Also, is there any doc of how to contribute to project?

**Comment by fehguy** [comment]
Hi,
Thanks for the suggestions!  As the project is Apache 2.0, you should submit code according to that license, and to the develop_2.0 branch.

For the packaging, I have some thoughts.
* I've felt that having samples in the same as the codegen project has made sense, since that ensures they stay in sync.
* swagger-parser is a separate project and is depended on by other projects outside the codegen project.  Since it's not going to be dependent on codegen, it seems like being separate is good.  It also lets it evolve independently and more rapidly (it's quite smaller)
* I did think a swagger-cli project could make sense.  There are two main classes in codegen library

Given the timing on the 2.1 release, I prefer to not repackage right now.

One thing that I really don't like is how the samples don't have native tests.  I'd love to get help on something like such:
* having a maven profile to rebuild the samples
* In the maven profile, we prop-up a stand-alone server for each language to run tests against
* in the maven profile, running tests under each sample to test the integrity of the generated code.  This would have to detect the available executables such as python, php, xcode, etc.

That would give much more confidence in the PRs.  It is complicated because it requires a bit of setup but I think worth the effort.

What do you think?

**Comment by lanwen** [comment]
Think we can stay in sync with help of CI, and there is no need to store samples in same repo for purpose of sync. I can help with jenkins setup for creating the build pipeline witch can test, release and deploy all components. Do you have some kind of ci.swagger.io, or only Travis build system?

I can setup jenkins for my local forks of swagger and then show it to you if no any CI enabled for now.

About CLI - think it same layer of abstraction as the web-app and should not be mixed with interface implementation. So we can use only interfaces in other apps such as `swagger-codegen-maven-plugin`, gradle plugin, servlet-only app, jersey app and so on, without need of cli importing. Also it will be not so difficult to understand where is the starting point of code generation.
Maybe it should not be separate repo, but it can be one of the separate modules.

I think good examples of code separation can be found here:
- [Allure Framework](https://github.com/allure-framework)
- [Jsonscheme2pojo](https://github.com/joelittlejohn/jsonschema2pojo)

**Comment by olensmar** [comment]
Hi!

my 5 cents: Relying on CI to stay in sync doesn’t seem right to me – wouldn’t it be better to have this at "code level”? Or am I old-fashioned? :-)

I would prefer separate modules to separate repos for the cli vs interfaces – I like to keep things together, it makes dependencies more obvious and understandable when you for example check out a single repo from github.

/Ole

**Comment by lanwen** [comment]
I think auto-generated stuff should be not commited to repo as it not written by human. We only need to test the process of generation and the final result. 
To stay in touch with samples it should not be updatable by human because of it's not very interesting - to launch every lang generator and then commit code which not understandable in most of cases by most of contributors (because of they fix only one or two langs and not all of them).

Usually, to see what is generated for specific language its easy to follow link to separate repo and do some clicks there. It's no need to download all of the samples to fix template of one of them. So - the latest version of the sample can be generated locally when developing fix. Other can be regenerated and tested on CI, and then, if all ok, autocommited to separate repo.

**Comment by fehguy** [comment]
Hi @lanwen, first thanks for the PRs thus for.  For the samples, I do have a concern with separating them from the main project.

In the absence of a better testing mechanism, I'm moving all samples to have unit tests against the petstore API.  The generated code should then pass in making calls to that API, using the vernacular of the target language.  So for java there are JUnit tests, for objective-c there are XCTest classes.  My expectation will be that before any merges for template or language-specific changes, the developer will run that integration test before it can be accepted.

If we separate these unit tests from the code generation logic, we'll end up writing invalid code.  Thus the two need to be more tightly coupled.

Do you have any better suggestion on how to test the output from the generated sources?  I don't particularily like this approach but the only other alternative I've seen is brittle too.

**Comment by lanwen** [comment]
CI can control process of testing all of generated code and thus don't mark as green the PR if any trouble.
Ability to generate and test all langs locally when develop is also should be, of course.

I'll setup the example in jenkins for this issue on next weekend, if i have enough time. I've already created special org for this purpose :)

**Comment by webron** [comment]
@lanwen - keep in mind we're not going to switch to jenkins at this point in time (perhaps in the future), so any useful set up would be only using travis.

**Comment by lanwen** [comment]
@webron why only travis?

**Comment by webron** [comment]
Because that's what we currently use for all the projects and there's no reason to complicate the procedure right now by adding another CI or switching to a different one (keep in mind we have quite a few projects and this is a maintenance issue). Not saying we won't consider moving to jenkins at some point, but so far Travis hasn't failed us (except for some speed issues which we seem to have resolved).

**Comment by fehguy** [comment]
To be fair, we are open to improving the CI process in the medium term as it makes sense.

**Comment by webron** [comment]
I agree, but I don't think we should split projects across CIs. It would end up being a headache to contributors and to us and should be planned well. If the same solution can be achieved using Travis, I can't think of a reason we'd need to migrate for this use case alone.

**Comment by fehguy** [comment]
So it looks like we're going to be moving the templates and samples into a different repo in 2.2.0.  Tagging this issue appropriately.
